signer/jws.go

```diff
@@ -0,0 +1,480 @@
+package signer
+
+import (
+	"crypto"
+	"crypto/ecdsa"
+	"crypto/rsa"
+	"crypto/x509"
+	"encoding/base64"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"reflect"
+	"strings"
+	"time"
+
+	"github.com/golang-jwt/jwt/v4"
+)
+
+const (
+	JwsJsonMediaType SignatureMediaType = "application/jose+json"
+)
+
+const (
+	headerKeyExpiry      = "io.cncf.notary.expiry"
+	headerKeySigningTime = "io.cncf.notary.signingTime"
+	headerKeyCrit        = "crit"
+	headerKeyAlg         = "alg"
+	headerKeyCty         = "cty"
+)
+
+// jwsInternalEnvelope represents the jwsEnvelope-JSON envelope
+type jwsEnvelope struct {
+	internalEnv *jwsInternalEnvelope
+}
+
+func newJWSEnvelopeFromBytes(envelopeBytes []byte) (*jwsEnvelope, error) {
+	jwsInternal, err := newJwsInternalEnvelopeFromBytes(envelopeBytes)
+	if err != nil {
+		return nil, err
+	}
+
+	return &jwsEnvelope{internalEnv: jwsInternal}, nil
+}
+
+func newJWSEnvelope() *jwsEnvelope {
+	return &jwsEnvelope{}
+}
+
+func (jws *jwsEnvelope) validateIntegrity() error {
+	if jws.internalEnv == nil {
+		return SignatureNotFoundError{}
+	}
+
+	if len(jws.internalEnv.Header.CertChain) == 0 {
+		return MalformedSignatureError{msg:"malformed leaf certificate"}
+	}
+
+	cert, err := x509.ParseCertificate(jws.internalEnv.Header.CertChain[0])
+	if err != nil {
+		return MalformedSignatureError{msg:"malformed leaf certificate"}
+	}
+
+	// verify JWT
+	compact := strings.Join([]string{jws.internalEnv.Protected, jws.internalEnv.Payload, jws.internalEnv.Signature}, ".")
+	return verifyJWT(compact, cert.PublicKey)
+}
+
+func (jws *jwsEnvelope) signPayload(req SignRequest) ([]byte, error) {
+	leafPublicKey := req.CertificateChain[0].PublicKey
+	var m map[string]interface{}
+	if err := json.Unmarshal(req.Payload, &m); err != nil {
+		return nil, err
+	}
+	signingMethod, err := getSigningMethod(leafPublicKey)
+	if err != nil {
+		return nil, err
+	}
+
+	signedAttrs := getSignedAttrs(req, signingMethod)
+	compact, err := signJWT(m, signedAttrs, signingMethod, req.SignatureProvider)
```

Comment:
`req.Payload` should already be a properly formatted payload ready to sign, so we could avoid unmarshaling and decoding it into signed signed attribues + claims if we just call `signingMethodForSign{algo: method.Alg(), sigProvider: sigPro}.Sign(req.Payload, "DummyNotUsed")` and then concatenating the resulting signature with the payload.